File: swift/Sources/MemCardKit/Lang.swift
import Foundation

/// Язык приложения. Хранится в файле настроек рядом с папками.
public enum Lang: String, CaseIterable, Identifiable, Sendable {
    case en, ru, fr, de, ja, zh, pl
    public var id: String { rawValue }

    /// Название языка на нём самом - так его узнают, не зная текущего.
    public var label: String {
        switch self {
        case .en: "English"
        case .ru: "Русский"
        case .fr: "Français"
        case .de: "Deutsch"
        case .ja: "日本語"
        case .zh: "中文"
        case .pl: "Polski"
        }
    }
}

/// Строки интерфейса.
///
/// **Ключ - английская строка.** Английский здесь исходный: он написан
/// прямо в коде, остальные языки лежат таблицами в `Resources/i18n`.
/// Нет перевода - показывается ключ, то есть английский, а не пустое
/// место и не имя переменной.
///
/// Своя таблица, а не `.strings` из бандла: у приложения, собранного
/// пакетом SPM, ресурсы лежат в отдельном бандле, и стандартная
/// локализация до них не дотягивается. Заодно язык переключается на
/// лету, без перезапуска.
///
/// **Не привязан к главному потоку намеренно.** Разбор сейвов идёт в
/// фоновой задаче, и подписи полей собираются там же - с `@MainActor`
/// они были бы недоступны.
public enum L {
    private static let lock = NSLock()
    nonisolated(unsafe) private static var value: Lang = .en
    nonisolated(unsafe) private static var table: [String: String] = [:]
    nonisolated(unsafe) private static var plurals: [String: [String: String]] = [:]

    public static var current: Lang {
        get { lock.lock(); defer { lock.unlock() }; return value }
        set {
            let (words, forms) = L.load(newValue)
            lock.lock()
            value = newValue
            table = words
            plurals = forms
            lock.unlock()
        }
    }

    /// Строка по ключу. Аргументы подставляются вместо `{0}`, `{1}` и так
    /// далее - порядок в переводе может быть любым, языки строят фразу
    /// по-разному.
    public static func t(_ key: String, _ args: Any...) -> String {
        lock.lock()
        var out = table[key]
        lock.unlock()
        // Одинаковые английские слова с разным смыслом различаются
        // пометкой после `@@`: «Save@@verb» - кнопка, «Save» - сейв.
        // Наружу пометка не идёт.
        if out == nil { out = key.components(separatedBy: "@@")[0] }
        var text = out!
        for (index, value) in args.enumerated() {
            text = text.replacingOccurrences(of: "{\(index)}", with: String(describing: value))
        }
        if text.contains("{{") || text.contains("}}") {
            text = text.replacingOccurrences(of: "{{", with: "{")
                       .replacingOccurrences(of: "}}", with: "}")
        }
        return text
    }

    /// Слово при числе. У русского и польского три формы, у французского
    /// и немецкого две, у японского и китайского одна - выбирает язык,
    /// а не место вызова.
    public static func plural(_ noun: String, _ count: Int) -> String {
        lock.lock()
        let forms = plurals[noun]
        let code = value
        lock.unlock()
        guard let forms else { return noun }
        return forms[L.form(code, count)] ?? forms["other"] ?? noun
    }

    private static func form(_ code: Lang, _ count: Int) -> String {
        switch code {
        case .ja, .zh:
            return "other"
        case .ru, .pl:
            let ten = count % 10, hundred = count % 100
            if ten == 1 && hundred != 11 { return "one" }
            if (2...4).contains(ten) && !(12...14).contains(hundred) { return "few" }
            return "many"
        case .fr:
            return count < 2 ? "one" : "other"
        case .en, .de:
            return count == 1 ? "one" : "other"
        }
    }

    private static func load(_ code: Lang) -> ([String: String], [String: [String: String]]) {
        guard let url = Bundle.module.url(forResource: "Resources/i18n/\(code.rawValue)",
                                          withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let raw = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return ([:], [:]) }
        var words: [String: String] = [:]
        var forms: [String: [String: String]] = [:]
        for (key, value) in raw {
            if let text = value as? String {
                words[key] = text
            } else if let group = value as? [String: String], key.hasPrefix("plural:") {
                forms[String(key.dropFirst("plural:".count))] = group
            }
        }
        return (words, forms)
    }
}

File: qt/nsm/lang.py
"""Строки интерфейса.

**Ключ - английская строка.** Английский здесь исходный: он написан прямо
в коде, остальные языки лежат таблицами в `tools/data/i18n`. Нет перевода -
показывается ключ, то есть английский, а не пустое место.

Таблицы общие с версией для macOS: там те же файлы лежат ресурсом внутри
приложения. Своя таблица, а не `gettext`, - чтобы язык переключался на
лету и не требовал сборки `.mo`.
"""

import json
import os

LANGUAGES = [("en", "English"), ("ru", "Русский"), ("fr", "Français"),
             ("de", "Deutsch"), ("ja", "日本語"), ("zh", "中文"),
             ("pl", "Polski")]

CODES = [code for code, _ in LANGUAGES]

CURRENT = "en"
_WORDS = {}
_PLURALS = {}

HERE = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
# Собранное приложение раскладывает `tools/data` ещё и рядом с собой.
DIRS = [os.path.join(HERE, "tools", "data", "i18n"),
        os.path.join(HERE, "data", "i18n"),
        os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data", "i18n")]


def _load(code):
    for base in DIRS:
        path = os.path.join(base, code + ".json")
        if os.path.exists(path):
            with open(path, encoding="utf-8") as handle:
                return json.load(handle)
    return {}


def set_language(code):
    global CURRENT, _WORDS, _PLURALS
    CURRENT = code if code in CODES else "en"
    raw = _load(CURRENT)
    _WORDS = {k: v for k, v in raw.items() if isinstance(v, str)}
    _PLURALS = {k[len("plural:"):]: v for k, v in raw.items()
                if k.startswith("plural:") and isinstance(v, dict)}


def t(key, *args):
    """Строка по ключу. Аргументы подставляются вместо `{0}`, `{1}` и так
    далее - порядок в переводе может быть любым."""
    text = _WORDS.get(key)
    if text is None:
        # Одинаковые английские слова с разным смыслом различаются пометкой
        # после `@@`: «Save@@verb» - кнопка, «Save» - сейв. Наружу она не идёт.
        text = key.split("@@")[0]
    if args:
        text = text.format(*args)
    elif "{{" in text or "}}" in text:
        text = text.replace("{{", "{").replace("}}", "}")
    return text


def plural(noun, count):
    """Слово при числе. Форму выбирает язык, а не место вызова: у русского
    и польского их три, у французского и немецкого две, у японского и
    китайского одна."""
    forms = _PLURALS.get(noun)
    if not forms:
        return noun
    return forms.get(_form(CURRENT, count)) or forms.get("other") or noun


def _form(code, count):
    if code in ("ja", "zh"):
        return "other"
    if code in ("ru", "pl"):
        ten, hundred = count % 10, count % 100
        if ten == 1 and hundred != 11:
            return "one"
        if 2 <= ten <= 4 and not 12 <= hundred <= 14:
            return "few"
        return "many"
    if code == "fr":
        return "one" if count < 2 else "other"
    return "one" if count == 1 else "other"


set_language(CURRENT)
